[HGI-306] - Add new error class and introduce error codes enum

varadarajan-tw opened the pull request:
In an effort to improve error handling and traceability for HGI destinations (more info [here](https://docs.google.com/document/d/1Q18pNSFVDvuvqacHg2y7CCoO1T2spaKl5To5Bzefuqw/edit#)), I did an audit on errors in the codebase, classified them to various categories and introduced new error classes for builders to build.

To bring in consistency to error codes and enable faster debugging, this PR adds:

- `PayloadValidationError` - Indicates error with fields in the payload. This can be used for validations performed within perform or performBatch. This 
- `Error Codes` - The `code` field of error is important as it shows up in Datadog and in Event Delivery tab. `Error Type` field is written to traces by `centrifuge` and `code` is set to  [`Error Type`](https://github.com/segmentio/integrations/blob/e833bc7508711d407ac7f95f7644f6c7af29dadf/Service/cloudevents.js#L873) field if not explicitly defined which happens in most of the cases. The enum values provide standard error code string for some of the common errors. 
- Updates some of the HGI/STRATCONN destinations that can use these new error codes.
- In addition to this, the descriptions for these new error codes have been added to Contentful. These descriptions should give additional context in debugging. Refer [this](https://docs.google.com/document/d/1BeMsDu19yCZdJ1WZzM31JadOpvdq2xyFyhrBeiJbuRU/edit) doc to view the new descriptions.

**Notes**:

- For AJVValidationErrors, the code field is set to 400 in [integrations](https://github.com/segmentio/integrations/blob/e833bc7508711d407ac7f95f7644f6c7af29dadf/createActionDestination/index.js#L316). If we can classify it as `INVALID PAYLOAD` error, then we will be able to clearly distinguish between segment errors and HttpErrors caused by partner APIs.  - Will verify if the assumption is true.
Here is the related PR - https://github.com/segmentio/integrations/pull/2521

## Testing

All unit tests are passing. There are no functional changes. Only the error codes have been refactored and these new codes should provide additional details for debugging due to to new descriptions set in [contentful](https://docs.google.com/document/d/1BeMsDu19yCZdJ1WZzM31JadOpvdq2xyFyhrBeiJbuRU/edit). Some of these errors are difficult to reproduce in staging. Here is a snapshot of `payload_validation_failed`.

<img width="1511" alt="image" src="https://user-images.githubusercontent.com/109586712/221552136-8f885c70-472d-4fd7-bf19-f326ff8a8c72.png">


- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

mericsson commented on packages/core/src/errors.ts:
```diff
@@ -62,9 +63,36 @@ export class RetryableError extends CustomError {
+ * Should include a user-friendly message.
+ * These errors will not be retried and the user has to fix the incorrect field mapping
+ */
+export class ValidationError extends IntegrationError {
```
Nice, great add!

mericsson commented on packages/core/src/errors.ts:
```diff
@@ -62,9 +63,36 @@ export class RetryableError extends CustomError {
+   * @param code - an optional error code/reason
+   * @param status - an optional http status code (e.g. 400)
+   */
+  constructor(message = '') {
```
Should we default an actual message like 'Validation error' ?

mericsson replied:
(Or is there a way in typescript to require uses of this to put a message? I'm unsure.)

varadarajan-tw replied:
Thanks for the comment. I removed the default empty string value. It is better to leave it as required param for the builder provide one.

rhall-twilio commented on packages/core/src/errors.ts:
```diff
@@ -62,9 +63,36 @@ export class RetryableError extends CustomError {
+ * Should include a user-friendly message.
+ * These errors will not be retried and the user has to fix the incorrect field mapping
+ */
+export class ValidationError extends IntegrationError {
```
Wonder if there's a way to pass a hint to the user of what field precisely failed validation.

varadarajan-tw replied:
Thanks for the comment Robert. Could you give an example on how the hints should be?
I added this class is to bring in consistency for `code` value and leave the error messages for builders to provide based on context.

rhall-twilio replied:
I believe the way the error message display works is a mapping of the error code. i.e, for `MISCONFIGURED_FIELD` it would show `Segment was unable to perform this request because a field in your settings is not configured correctly. Please revise them and try again` So the builder doesn't really get to pass out an error message that gets displayed in the UI, it only displays in the logs (someone can double check me?).

If so, it would be great if could allow the builder to pass `new ValidationError("The value of personas_computation_key and personas_audience_key must match.", "personas_computation_key, personas_audience_key")` so that the error in the UI can be upgraded to look like: 

`Segment was unable to perform this request because the field "personas_computation_key, personas_audience_key" in your settings is not configured correctly. Please revise them and try again`. 

I'm based this comment on a lot of assumptions about the Event Delivery errors, but at least that's what the _implementation looks like to me_. Think it's worth reviewing the implementation.

varadarajan-tw replied:
Thanks Robert! You are correct! The content for the generic description in the delivery overview is retrieved from [`contentful`](https://github.com/segmentio/gateway-api/blob/a023e5c619c0d278e119d4ffb03f16cf706adfb6/src/graphql/integrations/Integration.ts#L351). The customer given specific error messages can be only seen in `Response from Destination` tab. 
Some notes on Event Delivery [here](https://docs.google.com/document/d/1F2JxTZNk-MBmieCIe0wmvwdLr0n9yH2qMfcMBgAnMbM/edit#). (WIP) 

<img width="1508" alt="image" src="https://user-images.githubusercontent.com/109586712/217011581-c7b59a3d-33e6-4bd6-8713-eb5ac17909c6.png">

This means when we introduce new error codes, we also have to update the contentful with appropriate description to ensure proper error message is shown. Thanks for pointing this out.

On interpolating fields with contentful error messages: 
- We have to persist these fields in `clickhouse` (We will have to make changes to `integration`,  `centrifuge-visibility` and few other services in the persistence path)
- Alter the queries to retrieve the messages by errorType and fields 
- interpolate with contentful messages

This doesn't look straightforward and we can look at this option in the long term as it would require co-ordination across multiple teams. I ll this to our backlog.

varadarajan-tw replied:
I have added new descriptions to contentful. [This](https://docs.google.com/document/d/1BeMsDu19yCZdJ1WZzM31JadOpvdq2xyFyhrBeiJbuRU/edit) doc contains the new error codes and their description that has been set in contentful. For displaying hints to the user, i am working on a [change](https://github.com/segmentio/app/compare/error-handling?expand=1) in the app.

rhall-twilio commented on packages/destination-actions/src/destinations/google-analytics-4/addPaymentInfo/index.ts:
```diff
@@ -42,11 +42,7 @@ const action: ActionDefinition<Settings, Payload> = {
     // Google requires that `add_payment_info` events include an array of products: https://developers.google.com/analytics/devguides/collection/ga4/reference/events
     // This differs from the Segment spec, which doesn't require a products array: https://segment.com/docs/connections/spec/ecommerce/v2/#payment-info-entered
     if (payload.items && !payload.items.length) {
-      throw new IntegrationError(
```
I think it might be worth checking if we lose a lot of detail from the error message as we currently do not have MisconfiguredField in `contentful`. If that is the case, maybe we hold off merging and using the new error types until we populate some error message there.

varadarajan-tw replied:
Sure Robert. I am trying to get access to the right `contentful` project hosting these templates. On the other hand, even the current error codes do not have description set. For eg; `Misconfigured required field` shows up like this in event-delivery.
<img width="1509" alt="image" src="https://user-images.githubusercontent.com/109586712/217532260-cd9e4539-53fb-4cb7-8434-3d40bf7ac2e4.png">

I am trying to get access to the right project in contentful and I doubt the descriptions have been updated in sometime.

varadarajan-tw replied:
Marking this as resolved. Now that i have added Contentful messages, we should have additional information with these changes.

rhall-twilio commented:
We should update the README to provide new guidance on how to use these errors.

marinhero reviewed (commented):
Code looks good. Make sure to update the README and update the interface document with the guidelines that indicate how to use this.

brennan commented on packages/core/src/errors.ts:
```diff
@@ -62,9 +63,35 @@ export class RetryableError extends CustomError {
+/**
+ * Error to indicate the payload is missing fields that are required.
+ * Should include a user-friendly message.
+ * These errors will not be retried and the user has to fix the payload or .
```
```suggestion
 * These errors will not be retried and the user has to fix the payload.
```

brennan commented on packages/core/src/destination-kit/index.ts:
```diff
@@ -581,7 +581,7 @@ export class Destination<Settings = JSONObject> {
       const newTokens = await this.refreshAccessToken(destinationSettings, oauthSettings)
       if (!newTokens) {
-        throw new InvalidAuthenticationError('Failed to refresh access token')
+        throw new IntegrationError('Failed to refresh access token', ErrorCodes.OAUTH_REFRESH_FAILED)
```
When thinking about how to scale integrations moving forward, the more we can abstract away implementation from third-party builders, the better. So, I wonder if we should continue to use `InvalidAuthenticationError` here? This may be a good idea - because if we use the `IntegrationError` class, we would additionally need to pass in `401` as the status code.

varadarajan-tw replied:
Thanks for catching this. Fixed the issue by allowing users to set error code  for  `InvalidAuthenticationError`. The status code will remain `401`.

brennan commented on packages/core/src/errors.ts:
```diff
@@ -62,9 +63,35 @@ export class RetryableError extends CustomError {
 export class InvalidAuthenticationError extends CustomError {
   status = 401
-  code = 'invalid_authentication'
+  code = ErrorCodes.INVALID_AUTHENTICATION
```
We could do something like the below. This would allow us to continue using the `InvalidAuthenticationError` class for all of our authentication errors. This approach also makes it clear that all auth-related errors should have a status code of 401.

```suggestion
  code = this.code || ErrorCodes.INVALID_AUTHENTICATION
```

brennan commented on packages/core/src/errors.ts:
```diff
@@ -62,9 +63,35 @@ export class RetryableError extends CustomError {
 export class InvalidAuthenticationError extends CustomError {
   status = 401
-  code = 'invalid_authentication'
+  code = ErrorCodes.INVALID_AUTHENTICATION
```
We could do something like the below. This would allow us to continue using the `InvalidAuthenticationError` class for all of our authentication errors. This approach also makes it clear that all auth-related errors should have a status code of 401.

```suggestion
  code = this.code || ErrorCodes.INVALID_AUTHENTICATION
```

varadarajan-tw replied:
Made ErrorCodes.INVALID_AUTHENTICATION as the default value and it can be overriden through constructor param.

varadarajan-tw commented:
Thank you for all your suggestions - @marinhero , @rhall-twilio and @brennan . I have addressed the comments and also added an error handling guide as suggested.

brennan commented on docs/error-handling.md:
```diff
@@ -0,0 +1,46 @@
+
+All Http Errors resulting from API calls are automatically captured by the framework. The `request` object used for making HTTP API calls has an option called `throwHttpErrors`. In its default state of `true`, the `request` object throws an `HttpError` which are captured and handled automatically by the framework. Http Errors are retried depending on the status code. Refer [error.ts](../packages/ ...
+
+In case you wish to override this behavior, you can set `throwHttpErrors` explicitly to `false`. The `request` object would then not throw HttpError and return the `response` along with status code. You can then chose to throw your own version of the error or proceed with your open implementation
```
```suggestion
In case you wish to override this behavior, you can set `throwHttpErrors` explicitly to `false`. The `request` object would then not throw HttpError and return the `response` along with status code. You can then choose to throw your own version of the error or proceed with your open implementation.
```

brennan commented on docs/error-handling.md:
```diff
@@ -0,0 +1,46 @@
+
+## Custom error handling
+
+The inbuilt error handling should help you with most of the error scenarios you face. In cases where the validation capabilities provided by the framework is not sufficient or the error messages returned by the API you are interacting with are not helpful, you can chose to capture and throw your own custom errors. It is important to follow the following guidelines to ensure your errors are struct ...
```
```suggestion
The inbuilt error handling should help you with most of the error scenarios you face. In cases where the validation capabilities provided by the framework are not sufficient or the error messages returned by the API you are interacting with are not helpful, you can chose to capture and throw your own custom errors. It is important to adhere to the following guidelines to ensure your errors are structured properly and displayed to your destination users for appropriate action.
```

brennan commented on docs/error-handling.md:
```diff
@@ -0,0 +1,46 @@
+
+The inbuilt error handling should help you with most of the error scenarios you face. In cases where the validation capabilities provided by the framework is not sufficient or the error messages returned by the API you are interacting with are not helpful, you can chose to capture and throw your own custom errors. It is important to follow the following guidelines to ensure your errors are struct ...
+
+- DO NOT throw Javascript `Error` objects. Any error thrown from an action MUST contain a `message` describing the error, an `error code` indicating the type of error and a `status code` indicating the http status of the action. You MUST use predefined error classes defined in [error.ts](../packages/core/src/errors.ts). These classes help in capturing the necessary information in appropriate form ...
```
```suggestion
- DO NOT throw Javascript `Error` objects. Any error thrown from an action MUST contain a `message` describing the error, an `error code` indicating the type of error and a `status code` indicating the http status of the action. You MUST use predefined error classes defined in [error.ts](../packages/core/src/errors.ts). These classes help in capturing the necessary information in appropriate format. For example, assume that your action needs one of product id or product name. This kind of validation is not currently supported in `Action Definition`. Instead of `throw new Error('One of product id or name is required')`, use `throw new PayloadValidationError('One of product id or name is required')`. 
```

NolanChan reviewed (approved):
This is going to be a huge step towards better observability, thank you @varadarajan-tw !
